modules/post/windows/gather/enum_onedrive.rb

```diff
@@ -0,0 +1,176 @@
+##
+
+class MetasploitModule < Msf::Post
+  include Msf::Post::Windows::Priv
+  include Msf::Post::Common
+  include Msf::Post::File
+  include Msf::Post::Windows::Registry
+  include Msf::Post::Windows::UserProfiles
+
+  SYNC_ENGINES_KEYS = ['LibraryType', 'LastModifiedTime', 'MountPoint', 'UrlNamespace'].freeze
+  ONEDRIVE_ACCOUNT_KEYS = ['Business', 'ServiceEndpointUri', 'SPOResourceId', 'UserEmail', 'UserFolder', 'UserName'].freeze
+  PERSONAL_ONEDRIVE_KEYS = ['UserEmail', 'UserFolder'].freeze
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'OneDrive Sync Provider Enumeration Module',
+        'Description' => %q{
+          This module will identify the Office 365 OneDrive endpoints for both business and personal accounts
+          across all users (providing access is permitted). It is useful for identifying document libraries
+          that may otherwise not be obvious which could contain sensitive or useful information.
+        },
+        'License' => MSF_LICENSE,
+        'Platform' => ['win'],
+        'SessionTypes' => ['meterpreter'],
+        'Author' => ['Stuart Morgan <stuart.morgan[at]mwrinfosecurity.com>']
+      )
+    )
+  end
+
+  def display_report(sid, info, sync_used, sync_all, results_table)
+    info.each do |key, result|
+      next if result['ScopeIdToMountPointPathCache'].nil? || result['ScopeIdToMountPointPathCache'].empty?
+
+      row = []
+      print_line
+      print_line "  #{key}"
+      print_line "  #{'=' * key.length}"
+      print_line
+      row << sid
+      row << key
+      ONEDRIVE_ACCOUNT_KEYS.each do |col|
+        row << result[col].to_s
+        if result['Business'] == '1' || PERSONAL_ONEDRIVE_KEYS.include?(col)
+          print_line "    #{col}: #{result[col]}"
+        end
+      end
+      result['ScopeIdToMountPointPathCache'].each do |scopes|
+        subrow = row.clone
+        print_line
+        SYNC_ENGINES_KEYS.each do |sync|
+          subrow << scopes[sync].to_s
+          print_line "    | #{sync}: #{scopes[sync]}"
+        end
+        results_table << subrow
+      end
+    end
+
+    sync_all_list = []
+    sync_all.each do |key, _result|
+      sync_all_list.push(key)
+    end
+
+    diff = sync_all_list - sync_used
+    if !(diff.nil? || diff.empty?)
+      print_line
+      print_line '  ORPHANED'
+      print_line '  ========'
+      diff.each do |scopeid|
+        csvrow = []
+        print_line
+        # Augment the CSV
+        csvrow << sid
+        csvrow << ''
+        ONEDRIVE_ACCOUNT_KEYS.each do |_od|
+          csvrow << ''
+        end
+        SYNC_ENGINES_KEYS.each do |sync|
+          csvrow << sync_all[scopeid][sync]
+          print_line "  #{sync}: #{sync_all[scopeid][sync]}"
+        end
+        results_table << csvrow
+      end
+    end
+  end
+
+  def get_syncengine_data(master, syncengines)
+    all_syncengines = {}
+    syncengines.each do |ses|
+      newses = {}
```

Comment:
Okay so looks like in my run `ses` was `Buisness1` and then `Personal`, some form a hash, and then `Personal` again. I imagine it might be better to call these the account types rather than `ses` which sounds like session. Same goes for `newses`. Will update these names in the next commit.